internal/manager/run.go

```diff
@@ -180,7 +185,15 @@ func Run(
 	referenceIndexers := ctrlref.NewCacheIndexers(setupLog.WithName("reference-indexers"))
 	cache := store.NewCacheStores()
 	storer := store.New(cache, c.IngressClassName, logger)
-	configTranslator, err := translator.NewTranslator(logger, storer, c.KongWorkspace, translatorFeatureFlags)
+	configTranslator, err := translator.NewTranslator(logger, storer, c.KongWorkspace, translatorFeatureFlags, func() kong.AbstractSchemaService {
+		// returns schema service of an available gateway admin API client if there are any.
+		clients := clientsManager.GatewayClients()
+		if len(clients) > 0 {
+			return clients[0].AdminAPIClient().Schemas
+		}
+		// returns a fake schema service when no gateway clients available.
+		return translator.UnavailableSchemaService{}
+	})
```

Comment:
I'm not particularly fond of piggy backing the functional option here in order to reach out to enclosing scope for the `clientsManager`. How about instead we create a proper new function parameter, e.g. `SchemaServiceGetter`  (or something else, names are hard as always) and implement what is enclosed in lines 189-196 in that type's `GetSchemaService()`?

Then we could also place the related `UnavailableSchemaService` in that package/file and define mock/fake implementations for tests.

Some pseudo code:

```
type GatewayClientsProvider interface {
	GatewayClients() []*adminapi.Client
}

type SchemaServiceGetter struct {
  clientsManager GatewayClientsProvider
}

func NewSchemaServiceGetter(cm GatewayClientsProvider) SchemaServiceGetter {
  return SchemaServiceGetter}
    clientsManager: cm,
  }
}

func (ssg SchemaServiceGetter) GetSchemaService() kong.AbstractSchemaService {
		clients := ssg.clientsManager.GatewayClients()
		if len(clients) > 0 {
			return clients[0].AdminAPIClient().Schemas
		}
		// returns a fake schema service when no gateway clients available.
		return translator.UnavailableSchemaService{}
}
```

and then `NewTranslator` can accept `SchemaServiceGetter`.

WDYT?

Reply:
I can even see the interface already defined in https://github.com/Kong/kubernetes-ingress-controller/pull/6055/files#diff-4858537a492e6ebf6ced10c7e41b7bafa483f2ac870c8ae2854402aebeb36f13R133-R137

Reply:
Done.